src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp

```diff
@@ -481,11 +477,52 @@ EXTERN_C intptr_t RhGetCurrentThunkContext()
 //  thread        - thread to attach
 extern "C" void PalAttachThread(void* thread)
 {
-#if HAVE_THREAD_LOCAL
     tls_destructionMonitor.SetThread(thread);
-#else
-    __cxa_thread_atexit(RuntimeThreadShutdown, thread, &__dso_handle);
-#endif
+}
+
+struct Link
+{
+    Link* Next;
+    Destructor Dtor;
+    void* Arg;
+};
+
+static pthread_key_t key;
+static pthread_once_t once = PTHREAD_ONCE_INIT;
+
+static void CallDestructors(void* obj)
+{
+    Link* next;
+    Link* link = (Link*)obj;
+    while (link)
+    {
+        next = link->Next;
+        link->Dtor(link->Arg);
+        free(link);
+        link = next;
+    }
+}
+
+static void CreateKey(void)
+{
+    pthread_key_create(&key, CallDestructors);
+}
+
+int __cxa_thread_atexit(Destructor dtor, void* arg, void*)
```

Comment:
We should not need general purpose `__cxa_thread_atexit`. This can simplified to call just the thread destructor.

Reply:
clang gives an error:

```
  crossgen2 -> /runtime/artifacts/bin/coreclr/Linux.x64.Debug/crossgen2/linux-x64/crossgen2.dll
  Generating native code
  /usr/bin/ld: /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x25
  /runtime/artifacts/bin/coreclr/Linux.x64.Debug/aotsdk/libRuntime.ServerGC.a(PalRedhawkUnix.cpp.o): in function `__cxx_global_var_init':
  PalRedhawkUnix.cpp:(.text.startup+0x26): undefined reference to `__cxa_thread_atexit'
```

I did not find any way to satisfy the compiler without exporting `__cxa_thread_atexit` symbol. (Of course if we switch to pure C and use TLS in a C way, then these __cxa symbols are not needed)

Reply:
`__cxa_thread_atexit` is required by `thread_local TlsDestructionMonitor` that has a destructor. The only purpose of `TlsDestructionMonitor` is to notify the runtime about thread shutdown. If you delete `TlsDestructionMonitor` and change the pthread callback to notify the runtime directly, `__cxa_thread_atexit` should not be needed.

Reply:
I do not think that any of this requires switching to pure C.

Reply:
> `__cxa_thread_atexit` is required by `thread_local TlsDestructionMonitor` that has a destructor.

It gave me segmentation fault. The compiler is either indiscriminately calling `__cxa_thread_atexit` or there is more to `thread_local` which forces the compiler to go in the "MT mode" and it generates `__cxa_thread_atexit` call.

Reply:
That does not sound right. Where did the segmentation fault came from?

Reply:
```
  Generating native code
  crossgen2 -> /Users/adeel/projects/runtime/artifacts/bin/coreclr/OSX.arm64.Debug/crossgen2/osx-arm64/publish/
  Emitting R2R PE file: /Users/adeel/projects/runtime/artifacts/obj/Microsoft.NETCore.App.Crossgen2/Debug/net7.0/osx-arm64/S.P.C.tmp
  /var/folders/_p/q6xw9h591kx26frkrng5bfmw0000gn/T/tmp24e908c43c4e47ec9d0bbb953c01ae34.exec.cmd: line 2: 12772 Trace/BPT trap: 5       /Users/adeel/projects/runtime/artifacts/bin/coreclr/OSX.arm64.Debug/crossgen2/osx-arm64/publish/crossgen2 /Users/adeel/projects/runtime/artifacts/bin/coreclr/OSX.arm64.Debug/IL/System.Private.CoreLib.dll --out /Users/adeel/projects/runtime/artifacts/obj/Microsoft.NETCore.App.Crossgen2/Debug/net7.0/osx-arm64/S.P.C.tmp
/Users/adeel/projects/runtime/src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj(70,5): error MSB3073: The command "/Users/adeel/projects/runtime/artifacts/bin/coreclr/OSX.arm64.Debug/crossgen2/osx-arm64/publish/crossgen2 /Users/adeel/projects/runtime/artifacts/bin/coreclr/OSX.arm64.Debug/IL/System.Private.CoreLib.dll --out /Users/adeel/projects/runtime/artifacts/obj/Microsoft.NETCore.App.Crossgen2/Debug/net7.0/osx-arm64/S.P.C.tmp" exited with code 133.
```

It failed with SIGTRAP on osx-arm64 but worked on linux-x64. I can guard the `thread_local` code for non-linux.

Reply:
> I can guard the thread_local code for non-linux.

Sounds good to me. We have used pthread callbacks originally and switched to `thread_local` to fix macOS specific problems in https://github.com/dotnet/corert/pull/2531. We will need to stick with `thread_local` on macOS (or come up with some other way to fix the thread shutdown crash).